sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UnconditionalJumpStatementBase.cs

```diff
@@ -0,0 +1,149 @@
+
+        protected abstract DiagnosticDescriptor Rule { get; }
+        protected abstract GeneratedCodeRecognizer GeneratedCodeRecognizer { get; }
+
+        protected abstract LoopWalkerBase<TStatementSyntax, TLanguageKindEnum> GetWalker(SyntaxNodeAnalysisContext context);
+        protected abstract ISet<TLanguageKindEnum> LoopStatements { get; }
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(GeneratedCodeRecognizer,
+                c =>
+                {
+                    var walker = GetWalker(c);
+                    walker.Visit();
+                    foreach (var node in walker.GetRuleViolations())
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(Rule, node.GetLocation(), GetKeywordText(node)));
+                    }
+                },
+                LoopStatements.ToArray());
+        }
+
+        protected abstract string GetKeywordText(TStatementSyntax statement);
+    }
+
+    public abstract class LoopWalkerBase<TStatementSyntax, TLanguageKindEnum>
+        where TStatementSyntax : SyntaxNode
+        where TLanguageKindEnum : struct
+    {
+        protected readonly SyntaxNode rootExpression;
+
+        protected List<TStatementSyntax> ConditionalContinues { get; } = new List<TStatementSyntax>();
+        protected List<TStatementSyntax> ConditionalTerminates { get; } = new List<TStatementSyntax>();
+
+        protected List<TStatementSyntax> UnconditionalContinues { get; } = new List<TStatementSyntax>();
+        protected List<TStatementSyntax> UnconditionalTerminates { get; } = new List<TStatementSyntax>();
+
+        protected abstract ISet<TLanguageKindEnum> ConditionalStatements { get; }
+        protected abstract ISet<TLanguageKindEnum> StatementsThatCanThrow { get; }
+        protected abstract ISet<TLanguageKindEnum> LambdaSyntaxes { get; }
+
+        private ISet<TLanguageKindEnum> lambdaOrLoopStatements;
+        private ISet<TLanguageKindEnum> loopStatements;
+
+        public LoopWalkerBase(SyntaxNodeAnalysisContext context, ISet<TLanguageKindEnum> loopStatements)
+        {
+            rootExpression = context.Node;
+            this.loopStatements = loopStatements;
+            lambdaOrLoopStatements = LambdaSyntaxes.Union(loopStatements).ToHashSet();
+        }
+
+        public abstract void Visit();
+
+        public List<TStatementSyntax> GetRuleViolations()
+        {
+            var ruleViolations = new List<TStatementSyntax>(UnconditionalContinues);
+            if (!ConditionalContinues.Any())
+            {
+                ruleViolations.AddRange(UnconditionalTerminates);
+            }
+
+            return ruleViolations;
+        }
+
+        protected void StoreVisitData(TStatementSyntax node, List<TStatementSyntax> conditionalCollection,
+            List<TStatementSyntax> unconditionalCollection)

```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Consider using more general type 'System.Collections.Generic.ICollection<TStatementSyntax>' instead of 'System.Collections.Generic.List<TStatementSyntax>'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3242)